library/x509_create.c

```diff
@@ -123,53 +130,199 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name(const char *name,
     return cur;
 }
 
+static const x509_attr_descriptor_t *x509_attr_descr_from_numericoid(const char *numericoid,
+                                                                     size_t numericoid_len)
+{
+    const x509_attr_descriptor_t *cur;
+    mbedtls_asn1_buf *oid = mbedtls_calloc(1, sizeof(mbedtls_asn1_buf));
+    int ret;
+
+    ret = mbedtls_oid_from_numeric_string(oid, numericoid, numericoid_len);
+    if ((ret == MBEDTLS_ERR_X509_ALLOC_FAILED) || (ret == MBEDTLS_ERR_ASN1_INVALID_DATA)) {
+        mbedtls_free(oid->p);
+        mbedtls_free(oid);
+        return NULL;
+    }
+
+    for (cur = x509_attrs; cur->oid != NULL; cur++) {
+        if (strlen(cur->oid) == oid->len &&
+            strncmp(cur->oid, (const char *) oid->p, oid->len) == 0) {
+            break;
+        }
+    }
+
+    mbedtls_free(oid->p);
+    mbedtls_free(oid);
+    if (cur->oid == NULL) {
+        return NULL;
+    }
+
+    return cur;
+}
+
+static int hex_to_int(char c)
+{
+    return ('0' <= c && c <= '9') ? (c - '0') :
+           ('a' <= c && c <= 'f') ? (c - 'a' + 10) :
+           ('A' <= c && c <= 'F') ? (c - 'A' + 10) : -1;
+}
+
+static int hexpair_to_int(char c1, char c2)
+{
+    int n1 = hex_to_int(c1);
+    int n2 = hex_to_int(c2);
+    if (n1 != -1 && n2 != -1) {
+        return (n1 << 4) | n2;
+    } else {
+        return -1;
+    }
+}
+
+static int parse_attribute_value_string(const char *s,
+                                        int len,
+                                        unsigned char *data,
+                                        size_t *data_len)
+{
+    const char *c = s;
+    const char *end = c + len;
+    int hexpair = 0;
+    unsigned char *d = data;
+    int n;
+    while (c < end) {
+        if (*c == '\\') {
+            c++;
+
+            /* Check for valid escaped characters in RFC 4514 in Section 3*/
+            if (c + 1 < end && (n = hexpair_to_int(*c, *(c+1))) != -1) {
+                hexpair = 1;
+                *(d++) = n;
```

Comment:
For now I think we should explicitly reject escaped `'\0'` characters. Allowing the user to create certificates with strings that contain `'\0'` in the middle seems like a bad idea. UTF-8 work can address unacceptable characters more thoroughly, but it's best to have some safeguard.
```suggestion
                if (n == '\0') {
                    return MBEDTLS_ERR_X509_INVALID_NAME;
                }
                *(d++) = n;
```